### PR TITLE
Store configuration data in git

### PR DESCRIPTION
**Pull request by ljanyst**

### CODE REVIEW

File: cache/repo_cache.go [hunk 1]
```diff
@@ -1078,3 +1103,193 @@ func (c *RepoCache) finishIdentity(i *identity.Identity, metadata map[string]str
...
+
+		if entry.Name == "config.json" {
+			data, err := c.repo.ReadData(entry.Hash)
+			if err != nil {
+				return nil, fmt.Errorf("cache: failed to get data for blob %s (ref: %s): %s", entry.Hash, refName, err)
+			}
+			return data, nil
+		}
+	}
+	return nil, nil
```
Comment (dawzee) [inline cache/repo_cache.go:1224]: Should this not return an error? Config ref does not contain config.json blob

File: cache/repo_cache.go [hunk 2]
```diff
@@ -1078,3 +1103,193 @@ func (c *RepoCache) finishIdentity(i *identity.Identity, metadata map[string]str
...
+		exist, err := c.repo.RefExist(localRef)
+		if err != nil {
+			return fmt.Errorf("cache: failed to check if ref exists %s: %s", refName, err)
+		}
+
+		if !exist {
+			err = c.repo.CopyRef(remoteRef, localRef)
+			if err != nil {
+				return fmt.Errorf("cache: failed to copy %s to %s: %s", remoteRef, localRef, err)
+			}
```
Comment (dawzee) [inline cache/repo_cache.go:1251]: If there's no error then you could continue at this point

File: cache/repo_cache.go [hunk 3]
```diff
@@ -1078,3 +1103,193 @@ func (c *RepoCache) finishIdentity(i *identity.Identity, metadata map[string]str
...
+		}
+
+		localCommit, err := c.repo.ResolveRef(localRef)
+		if err != nil {
+			return fmt.Errorf("cache: failed to resolve %s: %s", localRef, err)
+		}
+		remoteCommit, _ := c.repo.ResolveRef(remoteRef)
```
Comment (dawzee) [inline cache/repo_cache.go:1258]: Should be "remoteCommit, **err** := c...."

File: cache/repo_cache.go [hunk 4]
```diff
@@ -774,6 +794,11 @@ func (c *RepoCache) Pull(remote string) error {
 		}
 	}
 
+	err = c.MergeConfigs(remote)
+	if err == nil {
```
Comment (dawzee) [inline cache/repo_cache.go:798]: err != nil?

File: cache/repo_cache.go (hunk 1 above)
Comment (ljanyst) [reply cache/repo_cache.go:1224]: Yes.

File: cache/repo_cache.go (hunk 2 above)
Comment (ljanyst) [reply cache/repo_cache.go:1251]: Good catch. It's not a problem though because of the equality check later on.

File: cache/repo_cache.go (hunk 4 above)
Comment (ljanyst) [reply cache/repo_cache.go:798]: Yes.

**Comment by ljanyst** [comment]
Thanks for the thorough review!

File: cache/repo_cache.go [hunk 5]
```diff
@@ -1078,3 +1103,200 @@ func (c *RepoCache) finishIdentity(i *identity.Identity, metadata map[string]str
...
+		if err != nil {
+			return fmt.Errorf("cache: failed to check if ref exists %s: %s", refName, err)
+		}
+
+		if !exist {
+			err = c.repo.CopyRef(remoteRef, localRef)
+			if err != nil {
+				return fmt.Errorf("cache: failed to copy %s to %s: %s", remoteRef, localRef, err)
+			}
+			return nil
```
Comment (dawzee) [inline cache/repo_cache.go:1258]: Suggest changing this to "continue" rather than "return nil" in future update in case there are multiple remoteRefs

**Pull request merged**